### PR TITLE
Vioapic: two minor refines

### PR DESCRIPTION
**Pull request by lifeix**
1. minor refine about ioapic madt parse
2. minor refine about vioapic_init

Tracked-On: #4550
Signed-off-by: Li Fei1 <fei1.li@intel.com>

### CODE REVIEW

**Comment by wenlingz** [comment]
OK to verify

**Comment by HuiHuangShi** [comment]
10 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:220<br> reason:Cast on a constant value.                    : madt<br><2> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:222<br> reason:Cast on a constant value.                    : madt<br><3> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:222<br> reason:For loop initialisation is not simple.<br><4> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:223<br> reason:Casting operation to a pointer. : (void* to struct acpi_subtable_header*): ( struct acpi_subtable_header * ) entry<br><5> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:223<br> reason:Casting operation on a pointer. : (void* to struct acpi_subtable_header*): ( struct acpi_subtable_header * ) entry<br><6> config:part_config filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:225<br> reason:Pointer not checked for null before use. : hdr<br><7> config:sharing_config filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:225<br> reason:Pointer not checked for null before use. : hdr<br><8> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:226<br> reason:Casting operation to a pointer. : (void* to struct acpi_madt_ioapic*): ( struct acpi_madt_ioapic * ) entry<br><9> config:common filename:/hypervisor/boot/acpi_base.c func:ioapic_parse_madt offset:226<br> reason:Casting operation on a pointer. : (void* to struct acpi_madt_ioapic*): ( struct acpi_madt_ioapic * ) entry<br><10> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:263<br> reason:Type conversion without cast. :  (struct acpi_table_madt* and void*): get_acpi_tbl<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: hypervisor/boot/acpi_base.c [hunk 1]
```diff
@@ -264,11 +257,10 @@ uint16_t parse_madt(uint32_t lapic_id_array[MAX_PCPU_NUM])
 uint8_t parse_madt_ioapic(struct ioapic_info *ioapic_id_array)
 {
 	uint8_t ret = 0U;
-	struct acpi_table_rsdp *rsdp = NULL;
+	const struct acpi_table_rsdp *rsdp = get_rsdp();
```
Comment (dongyaozu) [inline hypervisor/boot/acpi_base.c:260]: Typically, we only parse ACPI tables when CONFIG_ACPI_PARSE_ENABLED is defined.  Here it seems to be unconditionally.  Does it work when CONFIG_ACPI_PARSE_ENABLED is not defined?

File: hypervisor/boot/acpi_base.c [hunk 2]
```diff
@@ -210,27 +210,20 @@ local_parse_madt(struct acpi_table_madt *madt, uint32_t lapic_id_array[MAX_PCPU_
...
+	void *entry, *end;
+	struct acpi_subtable_header *hdr;
 	uint8_t ioapic_idx = 0U;
 
-	madt_ptr = (struct acpi_table_madt *)madt;
+	end = (void *)madt + madt->header.length;
 
-	first = madt_ptr + 1;
-	end = (void *)madt_ptr + madt_ptr->header.length;
+	for (entry = (void *)madt + sizeof(struct acpi_table_madt); (entry) < (end); entry += hdr->length) {
```
Comment (dongyaozu) [inline hypervisor/boot/acpi_base.c:222]: (void *)(madt + 1) ?

**Comment by HuiHuangShi** [comment]
14 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:235<br> reason:Type conversion without cast. :  (struct acpi_table_madt* and void*): get_acpi_tbl<br><2> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:241<br> reason:Cast from pointer to integral type. : (struct acpi_table_madt* to unsigned long): ( uint64_t ) madt<br><3> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:241<br> reason:Cast on a constant value.                    : madt<br><4> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:243<br> reason:Cast from pointer to integral type. : (struct acpi_table_madt* to unsigned long): ( uint64_t ) ( madt + 1 )<br><5> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:243<br> reason:For loop initialisation is not simple.<br><6> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:244<br> reason:Cast from integral type to pointer. : (unsigned long to Unknown_Type:acpi_subtable_header*): ( const struct acpi_subtable_header * ) entry<br><7> config:part_config filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:246<br> reason:Pointer not checked for null before use. : hdr<br><8> config:sharing_config filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:246<br> reason:Pointer not checked for null before use. : hdr<br><9> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:248<br> reason:Cast from integral type to pointer. : (unsigned long to Unknown_Type:acpi_madt_ioapic*): ( const struct acpi_madt_ioapic * ) entry<br><10> config:part_config filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:249<br> reason:Pointer not checked for null before use. : ioapic<br><11> config:sharing_config filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:249<br> reason:Pointer not checked for null before use. : ioapic<br><12> config:common filename:/hypervisor/dm/vioapic.c func:vioapic_indirect_write offset:288<br> reason:Narrower int conversion without cast. :  (unsigned char and unsigned int): vioapic -> chipinfo.id = data<br><13> config:common filename:/hypervisor/dm/vioapic.c func:vioapic_indirect_write offset:288<br> reason:Implicit conversion of underlying type (MR). :  (unsigned char and unsigned int): vioapic -> chipinfo.id = data<br><14> config:common filename:/hypervisor/dm/vioapic.c func:vioapic_init offset:554<br> reason:Expression is not Boolean.<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by HuiHuangShi** [comment]
13 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:238<br> reason:Type conversion without cast. :  (struct acpi_table_madt* and void*): get_acpi_tbl<br><2> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:241<br> reason:Cast from pointer to integral type. : (struct acpi_table_madt* to unsigned long): ( uint64_t ) madt<br><3> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:241<br> reason:Cast on a constant value.                    : madt<br><4> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:243<br> reason:Cast from pointer to integral type. : (struct acpi_table_madt* to unsigned long): ( uint64_t ) ( madt + 1 )<br><5> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:243<br> reason:For loop initialisation is not simple.<br><6> config:common filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:244<br> reason:Cast from integral type to pointer. : (unsigned long to Unknown_Type:acpi_madt_ioapic*): ( const struct acpi_madt_ioapic * ) entry<br><7> config:sharing_config filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:246<br> reason:Pointer not checked for null before use. : ioapic<br><8> config:part_config filename:/hypervisor/boot/acpi_base.c func:parse_madt_ioapic offset:246<br> reason:Pointer not checked for null before use. : ioapic<br><9> config:common filename:/hypervisor/dm/vioapic.c func:vioapic_indirect_write offset:288<br> reason:Implicit conversion of underlying type (MR). :  (unsigned char and unsigned int): vioapic -> chipinfo.id = data<br><10> config:common filename:/hypervisor/dm/vioapic.c func:vioapic_indirect_write offset:288<br> reason:Narrower int conversion without cast. :  (unsigned char and unsigned int): vioapic -> chipinfo.id = data<br><11> config:sharing_config filename:/hypervisor/dm/vioapic.c func:vioapic_init offset:524<br> reason:Attempt to use uninitialised pointer. : vioapic_info<br><12> config:part_config filename:/hypervisor/dm/vioapic.c func:vioapic_init offset:524<br> reason:Attempt to use uninitialised pointer. : vioapic_info<br><13> config:common filename:/hypervisor/dm/vioapic.c func:vioapic_init offset:552<br> reason:Expression is not Boolean.<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by wenlingz** [comment]
OK to verify

**Comment by wenlingz** [comment]
OK to verify

**Pull request merged**